docs/proposals/pod-safety.md

```diff
@@ -0,0 +1,388 @@
+one pod instance is running on the cluster at any time for a given pod name,
+it must be possible to make pod deletion strongly consistent.
+
+To do that, we will:
+
+* Give the Kubelet sole responsibility for normal deletion of pods -
+  only the Kubelet in the course of normal operation should ever remove a
+  pod from etcd (only the Kubelet should force delete)
+  * The kubelet must not delete the pod until all processes are confirmed
+    terminated.
+* Application owners must be free to force delete pods, but they *must*
+  understand the implications of doing so, and all client UI must be able
+  to communicate those implications.
+* All existing controllers in the system must be limited signaling pod
+  termination (starting graceful deletion), and are not allowed to force
+  delete a pod.
+  * The node controller will no longer be allowed to force delete pods -
+    it may only signal deletion
+  * The GC controller may not force delete pods
+  * The namespace controller used to force delete pods, but no longer
+    does so. This means a node partition can block namespace deletion
+    indefinitely.
+  * The pod GC controller may continue to force delete pods on nodes that
+    no longer exist if we treat node deletion as confirming permanent
+    partition. If we do not, the pod GC controller must not force delete
+    pods.
+* It must be possible for an administrator to effectively resolve partitions
+  manually to allow namespace deletion.
+* Deleting a node from etcd can be seen as a signal to the cluster that
+  the node is permanently partitioned. We must audit existing components
+  to verify this is the case.
+  * Alternatively, we could require that pods must be individually
+    terminated. Complicates cloud controller interactions, but may be
+    easier for admins to reason about.
+
+In the above scheme, force deleting a pod releases the lock on that pod and
+allows higher level components to proceed to create a replacement.
+
+It has been requested that force deletion be restricted to privileged users.
+That limits the application owner in resolving partitions when the consequences
+of force deletion are understood, and not all application owners will be
+privileged users. For example, a user may be running a 3 node etcd cluster in a
+pet set. If pet 2 becomes partitioned, the user can instruct etcd to remove
+pet 2 from the cluster (via direct etcd membership calls), and because a quorum
+exists pets 0 and 1 can safely accept that action. The user can then force
+delete pet 2 and the pet set controller will be able to recreate that pet on
+another node and have it join the cluster safely (pets 0 and 1 constitute a
+quorum for membership change).
+
+
+### Fencing
+
+The changes above allow Pet Sets to ensure at-most-one pod, but provide no
+recourse for the automatic resolution of cluster partitions during normal
+operation. For that, we propose a **fencing controller** which exists above
```

Comment:
Forgiveness hasn't been fully implemented yet but I was assuming that once the forgiveness period expires, it would delete the same way node controller deletes today when there's a NotReady (bad node/kubelet) or Unknown (node unreachable), i.e. the code here
https://github.com/kubernetes/kubernetes/blob/4bda90e89c1e945c74856553b47a517a3244f874/pkg/controller/node/controller_utils.go#L45
which IIUC does a regular delete not a force delete? (Where is the client Delete() operation that it's using defined?)